feat(topology): Topology plugin visualisation of workload (#104)
janus-idp/backstage-plugins: plugins/topology/README.md

# Topology plugin for Backstage

> This plugin will help with visualizing the workloads like Deployment, Replicasets, Pods powering any serivce on kubernets cluster.

Reviewer: ```suggestion
This plugin will help with visualizing the workloads like Deployment, Replicaset and Pods powering any service on the Kubernetes cluster.
```

## Prerequisites

> 1. Install and configure kubernetes plugin by following these guides:

Reviewer: ```suggestion
1. Install and configure Kubernetes plugin by following these guides:
```

   https://backstage.io/docs/features/kubernetes/installation
   https://backstage.io/docs/features/kubernetes/configuration
2. Kubernetes plugin is properly configured and able to connect to the cluster via a `ServiceAccount`

> 3. Following `ClusterRole` must be granted to ServiceAccount accessing the cluster: (k8s plugin by default tries to fetch the following resources and if access is not provided an error is shown)

Reviewer 2: I think we should mention this that this role is already granted if you have the Backstage Kubernetes Plugin properly configured (which is a prerequisite), and link to https://backstage.io/docs/features/kubernetes/configuration#role-based-access-control

The current formulation gives impression that there is something different in the role itself when compared to the one required by kubernetes plugin. While that may become the case in future, I think we should rather formulate in a way that basically says: "Please grant the permissions that Backstage Kubernetes plugin requires + these additional ones" (once we have any additional ones).

   ```yaml
   annotations:
     backstage.io/kubernetes-id: <BACKSTAGE_ENTITY_NAME>
     backstage.io/kubernetes-namespace: <RESOURCE_NS>
   ```
5. A custom label selector can also be added which will then be used by Backstage to find the resources. The label selector takes precedence over the id annotation.
   `'backstage.io/kubernetes-label-selector': 'app=my-app,component=front-end`
6. For k8s plugin to get the desired entity's k8s resources the following label should be added to the resources(if label selector is used then the labels mentioned in that should be present on the resource):
   ```yaml
   'backstage.io/kubernetes-id': <BACKSTAGE_ENTITY_NAME>`
   ```
7. Workload resources like Deployments and Pods will be shown in a visual group based on below label
   ```yaml
   'app.kubernetes.io/part-of': <GROUP_NAME>
   ```
8. Workload resources like Deployment or Pod will be shown with a visual connector based on below annotations
   ```yaml
   'app.openshift.io/connects-to': '[{"apiVersion": <RESOURCE_APIVERSION>,"kind": <RESOURCE_KIND>,"name": <RESOURCE_NAME>}]'
   ```

## Getting started

1. Install the plugin

   ```bash

>    yarn workspace app add @janus-idp/backstage-plugin-topology

Reviewer: ```suggestion
   yarn add --cwd packages/app @janus-idp/backstage-plugin-topology
```

Reviewer 2: Is there a reason why this should be preferred? Can you please elaborate?

Reviewer: For most of the upstream packages for the backstage this format is commonly used to make sync on this I suggested

# Topology plugin for Backstage

This plugin will help with visualizing the workloads like Deployment, Replicaset and Pods powering any service on the Kubernetes cluster.

## Prerequisites

1. Install and configure Kubernetes plugin by following these guides:
   https://backstage.io/docs/features/kubernetes/installation
   https://backstage.io/docs/features/kubernetes/configuration

> 2. Kubernetes plugin is properly configured and able to connect to the cluster via a `ServiceAccount`

Author: It should be any supported authentication mechanism as provided here https://backstage.io/docs/features/kubernetes/configuration 

and has cluster roles for https://backstage.io/docs/features/kubernetes/configuration#role-based-access-control 

we have added `ServiceAccount` as that's what we have tested. Once we verify other mechanisms will update.

@divyanshiGupta ^^
